Material Compiler Merge necessary for Simple Raytracer

devshgraphicsprogramming opened the pull request.

Crisspl committed:
Envmap sampling works again

TODO correct plastic sampling and dielectric
next: BSDFs sampling

Crisspl committed:
Plastic sampling

Well, this looks bad, i think in the future plastic plastic (and coating as well) could be divided into smaller instructions and work like double-weighted blend

Crisspl committed:
Merge with devsh/shader_pipeline

Crisspl committed:
WIP changes and adjustments to mat compiler GLSL after API change

Crisspl committed:
mat compiler: Changes after API changes

Notes:
* im doing max(alpha,1e-4) to avoid alpha=0 because NDFs often result in NaNs then, this want problem a before when i computed remainder only
* dielectric is working, but thindielectric is black for some reason
* im putting 1.0 instead for PDFs which were INF

Crisspl committed:
mat compiler: Derivative maps finally works!

Crisspl committed:
A few adjustments for mat compiler due to API change

Crisspl committed:
Updated ex45 shader:

Crisspl committed:
Merge with devsh/shader_pipeline

Crisspl committed:
tmp

Crisspl committed:
Merge branch 'shader_pipeline' of github.com:buildaworldnet/IrrlichtBAW into mitsuba

Crisspl committed:
Crisis handled, pt1

Some TODOs left, but working

Crisspl committed:
Mat compiler: transmissive beckmann (OP_DIELECTRIC) is working

Crisspl committed:
Getting rid of OP_SET_GEOM_NORMAL at the beginning of streams

Crisspl committed:
Update ex18

Crisspl committed:
Merge branch 'shader_pipeline' of github.com:buildaworldnet/IrrlichtBAW into mitsuba

Crisspl committed:
mat compiler: Added transmissive GGX and adjusted shaders due to merge

Crisspl committed:
mat compiler: Fixed plastic generator choice to be weighted by fresnel(NdotV) for specular

Crisspl committed:
Fixed remainder_and_pdf for BRDFs (was zero pdf resulting in nans sometimes)

Conductors look nice again in material compiler, it was shit af

Crisspl committed:
Fixed thin dielectric

Crisspl committed:
A little trick for plastic sampling

A problem was when diffuse generator had been chosen and tried to evaluate beckmann specular with NdotH=0.0

Crisspl committed:
Changed this log level to ERROR

Crisspl committed:
Fixed a few missing things in material compiler

Crisspl committed:
Finalizing mitsuba loader/mat compiler raster backend

Crisspl committed:
Resolved a few missing things in mitsuba loader

Mostly:
* cond. optim param tex presence flags checking
* cond. optim parameters fetching if always returning same value
* hoisted cosFactor and reg accesses outside instr_execute_cos_eval_OPCODE()
I have no idea how instancing worked before (tahts why i did my change) and im unable to undo this (i tried and failed) -- mostly because what CElementShape::getAbsoluteTransform() does in case of INSTANCE type is concatenate local and instance's tforms. The thing is the child of instance node must be shapegroup node and shapegroups doesnt have transform by themselves. Thats why this code didnt make sense

Crisspl committed:
Fixed a few minor issues in mitsuba loader

Crisspl committed:
dont push this

Crisspl committed:
ITraveralGenerator typo fix

Crisspl committed:
Material compiler: use switch/case instead of if/else with a single #define

Crisspl committed:
Few small fixes in mitsuba loader/material compiler

Dervi map cache and instr streams change from 4x (offset,count) to offset + 4x count

Crisspl committed:
Make use of irr_glsl_conditionalAbsOrMax() in mat compiler

Crisspl committed:
plastic2coating wip

Crisspl committed:
mitsuba loader/mat compiler: Some basic coating, some fixes to twosided and more

Crisspl committed:
Lots of minor fixes in mitsuba loader and material compiler

Crisspl committed:
mat compiler: reworked prefetch stream, almost working

Crisspl committed:
Ok, fixed one bug, but there's more

At least staircase works fine, bathroom does not

Crisspl committed:
Fixed some overflows

Crisspl committed:
Fixed new prefetch stream and sobol seq generation

Crisspl committed:
Got rid of in-loader image cache, reworked image caching in mitsuba loader and more

Crisspl committed:
Normalized normal and view precomputation

Crisspl committed:
Dedicated IR node for dielectric; moved all backend-specific definitions into class definition, some minor fixes to GLSL

Crisspl committed:
Merge with 3rdparty-upgrade branch

Crisspl committed:
Some fixes after merge with 3rdparty-upgrade

Crisspl committed:
Use bsdf_id string in debug only

Crisspl committed:
mat compiler IR: free tmp nodes space when they're not needed any more

Crisspl committed:
Multiple fixes of comments from PR

Crisspl committed:
More robust check for generator instruction in rem_and_pdf stream execution

Crisspl committed:
Better generator handling code : WIP

Im going to transform COATING into BLEND frist, it will simplify some things

Crisspl committed:
Added sample validation

Crisspl committed:
Revert "Added sample validation"

This reverts commit eae45a9cbc772a88f90e7b6ac9eff370d61fbfaa.

Crisspl committed:
Changed coating into something blend-like (on backend-side)

Crisspl committed:
New generator handling code

It's broken though, i think it's because i use OP_DIELECTRIC as specular part of OP_COATING

devshgraphicsprogramming commented on include/irr/builtin/glsl/sampling/cos_weighted.glsl:
```diff
@@ -22,6 +22,11 @@ float irr_glsl_projected_hemisphere_remainder_and_pdf(out float pdf, in vec3 L)
 	return irr_glsl_projected_hemisphere_remainder_and_pdf(pdf,L.z);
 }
 
+float irr_glsl_projected_hemisphere_pdf(in float L_z)
```
I hate duplicate code @Crisspl make sure that `float irr_glsl_projected_hemisphere_remainder_and_pdf(out float pdf, in float L_z)` uses `float irr_glsl_projected_hemisphere_pdf(in float L_z)`

devshgraphicsprogramming replied:
duplicate code

devshgraphicsprogramming replied:
still duplicate code

devshgraphicsprogramming commented on include/irr/builtin/glsl/sampling/cos_weighted.glsl:
```diff
@@ -40,4 +45,9 @@ float irr_glsl_projected_sphere_remainder_and_pdf(out float pdf, in vec3 L)
     return irr_glsl_projected_sphere_remainder_and_pdf(pdf,L.z);
 }
 
+float irr_glsl_projected_sphere_pdf(in float L_z)
```
same thing as https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r518968449

devshgraphicsprogramming replied:
duplicate code

devshgraphicsprogramming replied:
still duplicate code... I only want one place where the following code exists `L_z * irr_glsl_RECIPROCAL_PI`

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/beckmann.glsl:
```diff
@@ -112,6 +112,10 @@ float irr_glsl_beckmann_pdf_wo_clamps(in float NdotH2, in float maxNdotV, in flo
+float irr_glsl_beckmann_pdf(in irr_glsl_IsotropicViewSurfaceInteraction i, irr_glsl_IsotropicMicrofacetCache h, in float a2)
+{
+    return irr_glsl_beckmann_pdf_wo_clamps(h.NdotH2, max(i.NdotV, 0.0), i.NdotV_squared, a2);
+}
```
why do these functions exist?

devshgraphicsprogramming replied:
?

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893462 + https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893512

they shouldn't exist as standalones

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/beckmann.glsl:
```diff
@@ -130,6 +134,16 @@ float irr_glsl_beckmann_pdf_wo_clamps(in float NdotH2, in float TdotH2, in float
+    float TdotV2 = i.TdotV * i.TdotV;
+    float BdotV2 = i.BdotV * i.BdotV;
+    return irr_glsl_beckmann_pdf_wo_clamps(h.isotropic.NdotH2, TdotH2, BdotH2, maxNdotV, TdotV2, BdotV2, NdotV2, ax, ax2, ay, ay2);
+}
```
why do these functions exist?

devshgraphicsprogramming replied:
?

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893462 + https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893512

they shouldn't exist as standalones

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/beckmann.glsl:
```diff
@@ -151,7 +165,7 @@ vec3 irr_glsl_beckmann_cos_remainder_and_pdf(out float pdf, in irr_glsl_LightSam
     else
     {
-        pdf = 0.0;
+        pdf = irr_glsl_beckmann_pdf(interaction, _cache, a2);
```
pdf should be 0.0, because either the viewer or light is behind the surface and this is a BRDF

devshgraphicsprogramming replied:
my bad, if it actually gets past an `is_Valid` check (the checks of microfacet orientation) for a BRDF...

then we need to compute the PDF.

BUT dont branch and compute the same PDF functions in both branches! UNIFY!

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/beckmann.glsl:
```diff
@@ -188,7 +202,7 @@ vec3 irr_glsl_beckmann_aniso_cos_remainder_and_pdf(out float pdf, in irr_glsl_Li
     else
     {
-        pdf = 0.0;
+        pdf = irr_glsl_beckmann_pdf(interaction, _cache, ax, ax*ax, ay, ay*ay);
```
pdf should be 0.0, because either the viewer or light is behind the surface and this is a BRDF

devshgraphicsprogramming replied:
my bad, if it actually gets past an `is_Valid` check (the checks of microfacet orientation) for a BRDF...

then we need to compute the PDF.

BUT dont branch and compute the same PDF functions in both branches! UNIFY!

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl:
```diff
@@ -142,6 +142,10 @@ float irr_glsl_ggx_pdf_wo_clamps(in float NdotH2, in float maxNdotV, in float Nd
+float irr_glsl_ggx_pdf(in irr_glsl_IsotropicViewSurfaceInteraction i, irr_glsl_IsotropicMicrofacetCache h, in float a2)
+{
+    return irr_glsl_ggx_pdf_wo_clamps(h.NdotH2, max(i.NdotV, 0.0), i.NdotV_squared, a2);
+}
```
same comment as with beckmann header

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893462 + https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893512

they shouldn't exist as standalones

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl:
```diff
@@ -150,7 +154,16 @@ float irr_glsl_ggx_pdf_wo_clamps(in float NdotH2, in float TdotH2, in float Bdot
+    float TdotV2 = i.TdotV*i.TdotV;
+    float BdotV2 = i.BdotV*i.BdotV;
+    return irr_glsl_ggx_pdf_wo_clamps(h.isotropic.NdotH2, TdotH2, BdotH2, maxNdotV, NdotV2, TdotV2, BdotV2, ax, ay, ax2, ay2);
+}
```
same comment as with beckmann header

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893462 + https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893512

they shouldn't exist as standalones

devshgraphicsprogramming replied:
TODO

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl:
```diff
@@ -175,7 +188,7 @@ vec3 irr_glsl_ggx_cos_remainder_and_pdf(out float pdf, in irr_glsl_LightSample _
     else
     {
-        pdf = 0.0;
+        pdf = irr_glsl_ggx_pdf(interaction,_cache,a2);
```
same comment as with beckmann header

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893462 + https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893512

devshgraphicsprogramming replied:
UNIFY

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl:
```diff
@@ -217,7 +230,7 @@ vec3 irr_glsl_ggx_aniso_cos_remainder_and_pdf(out float pdf, in irr_glsl_LightSa
     else
     {
-        pdf = 0.0;
+        pdf = irr_glsl_ggx_pdf(interaction, _cache, ax, ay, ax*ax, ay*ay);
```
same comment as with beckmann header

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893462 + https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527893512

devshgraphicsprogramming replied:
UNIFY

Crisspl committed:
Ok fixed coating sampling

Seems like dielectric is broken because it doesnt have reflection or transmission at all

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -1140,6 +1133,11 @@ irr_glsl_LightSample irr_bsdf_cos_generate(in MC_precomputed_t precomp, in instr
 		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+
+		const float alpha = is_bsdf ? 0.0 : ax2;
+		const vec3 reflectance = is_bsdf ? vec3(1.0) : refl;
```
whats the messing around with alpha and reflectance?

devshgraphicsprogramming replied:
why is the diffuse smooth when its a transmissive? and why is it not allowed an albedo?

Crisspl committed:
mat compiler: Fixes to sample generation and BLEND (vec3 weight instead of float)

Crisspl committed:
Optimized eval and eval_and_pdf streams execution for divergence

Got rid of black and white pixels as well

Crisspl committed:
Those can go

Crisspl committed:
mat compiler: Implementing opacity as blend with delta transmission

Crisspl committed:
Ehh stupid bug, opacity works now

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -1262,7 +1274,7 @@ vec3 runGenerateAndRemainderStream(in MC_precomputed_t precomp, in instr_stream_
 	out_pdf = pdf;
 
-	vec3 rem = generator_rem/(1.0 + restPdf/generator_pdf) + acc/pdf;
+	vec3 rem = (generator_rem + acc/generator_pdf) / (1.0 + restPdf/generator_pdf);
```
`1.0/generator_pdf` could be precomputed

Crisspl committed:
Made bsdf_data much smaller (got rid of opacity)

Crisspl committed:
WIP new twosided approach

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/bsdf/specular/beckmann.glsl:
```diff
@@ -190,4 +213,45 @@ float irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_wo_cache_validati
+        _cache.isotropic.NdotH2, TdotH2, BdotH2, _sample.NdotL2, TdotL2, BdotL2, absNdotV,
+        NdotV2, TdotV2, BdotV2, transmitted, VdotH, LdotH,
+        VdotHLdotH, orientedEta, orientedEta2, ax, ax2, ay, ay2
+    );
```
this is doing duplicate computation

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/bsdf/specular/beckmann.glsl:
```diff
@@ -145,6 +145,29 @@ float irr_glsl_beckmann_smith_height_correlated_dielectric_cos_eval_wo_cache_val
+
+    //(in bool transmitted, in float reflectance, in float ndf, in float absNdotV, in float NdotV2, in float VdotH, in float LdotH, in float VdotHLdotH, in float a2, in float orientedEta, out float onePlusLambda_V
+    pdf = irr_glsl_beckmann_dielectric_pdf_wo_clamps(transmitted, reflectance, ndf, absNdotV, NdotV2, _cache.VdotH, _cache.LdotH, VdotHLdotH, a2, orientedEta, dummy);
+    return irr_glsl_beckmann_smith_height_correlated_dielectric_cos_eval_wo_clamps(NdotH2, NdotL2, absNdotV, NdotV2, transmitted, _cache.VdotH, _cache.LdotH, VdotHLdotH, orientedEta, orientedEta2, a2);
```
duplicate computation

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/common.glsl:
```diff
@@ -70,6 +68,12 @@ irr_glsl_LightSample irr_glsl_createLightSampleTangentSpace(in vec3 tangentSpace
+{
+    vec3 tangentSpaceL = irr_glsl_reflect(tangentSpaceV, tangentSpaceH, VdotH);
+    return irr_glsl_createLightSampleTangentSpace(tangentSpaceV, tangentSpaceL, tangentFrame);
+}
```
why do you need this?

Crisspl replied:
now i dont, not used

devshgraphicsprogramming replied:
delete it then

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/common.glsl:
```diff
@@ -108,6 +112,11 @@ irr_glsl_LightSample irr_glsl_createLightSample(in vec3 L, in irr_glsl_Anisotrop
     return irr_glsl_createLightSample(L,dot(interaction.isotropic.V.dir,L),interaction.T,interaction.B,interaction.isotropic.N);
 }
 
+vec3 irr_glsl_getTangentSpaceL(in irr_glsl_LightSample s)
+{
+    return vec3(s.TdotL, s.BdotL, s.NdotL);
+}
```
where is this used?

Crisspl replied:
nowhere, just a utility

devshgraphicsprogramming replied:
delete it them

Crisspl replied:
ok.... why? XD

devshgraphicsprogramming replied:
if we dont use something no point keeping it around (we have git history after all)

devshgraphicsprogramming replied:
?

devshgraphicsprogramming replied:
why do we still have it?

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
 	uint ndf = instr_getNDF(instr);
 	float a = params_getAlpha(params);
 	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
```
what happened to this optimization?

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
-		if (ndf==NDF_GGX) {
+#ifndef NO_TWOSIDED
+	handleTwosided(instr, s, _microfacet);
 #endif
```
is the proper twosided handling implemented yet? because with different instruction streams per face/side this function shouldn't exist yet

Crisspl replied:
thats why i didnt even bother with proper twosided handling

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
-#else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
```
move this and the test for `if (cosFactor>FLT_MIN)` to the very top

devshgraphicsprogramming replied:
will save you a lot of data fetches and logic if you have a BRDF or a very "on edge" BSDF

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
```
after you do twosided properly, there will be no more need to orient the Etas

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
+		if (op_isDiffuse(op))
+		{
+			vec3 reflectance = is_bsdf ? vec3(1.0) : refl;
+			float alpha2 = is_bsdf ? 0.0 : a2;
```
still, whats with the reflectance and roughness modification?

Crisspl replied:
this vec3(1.0) is wrong, changed it to transmittance. alpha2==0.0 in case of bsdf is because OP_DIFFTRANS is always smooth (doesnt even have roughness parameter)

devshgraphicsprogramming replied:
> because OP_DIFFTRANS is always smooth (doesnt even have roughness parameter)

we could allow a rough DIFFTRANS (in the IR) at no extra cost (excution or complexity)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
+				float n = irr_glsl_alpha2_to_phong_exp(a2);
+				float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+				bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(s, microfacet, currInteraction, n, ny, a2, ay2);
+			} CASE_END
```
if transmission occurs, how will the `cos_eval_DG` functions play with the fact that one of `VdotN` or `LdotN` will be negative? especially the smith masking function?

Furthermore at least one of `VdotH` and `LdotH` will be negative too!

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,189 @@
-		instr_execute_cos_eval_BLEND(instr, regs, params, bsdf_data);
-	} else
+		CASE_BEGIN(op, OP_BLEND) {
+			result = instr_execute_cos_eval_BLEND(instr, srcs, params, bsdf_data);
```
so coating is still not unified with blend?

Crisspl replied:
it is blend-like, but has different opcode. Honestly i cant see any point in unifying them further, in the end of the day they differ a lot (first of all blend takes weight from bsdf_data or texture and coating computes it). So there would be a need for flag determining how weights are obtained which already stands for divergence between most of the work done by the opcode

devshgraphicsprogramming replied:
reading and writing data to random registers might be expensive... so it would be nice not to diverge for that
(which is why single opcode is nice)

devshgraphicsprogramming replied:
hmm ok I'll allow a COAT which is a different opcode to BLEND

devshgraphicsprogramming replied:
ok I guess its fine

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -735,179 +734,307 @@ void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
+	if (thinOrDelta)
+	{
+		bxdf_eval = vec3(0.0);
+		pdf = 1.0;
```
wrong

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -735,179 +734,307 @@ void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
+	if (op_isDiffuse(op))
+	{
+		vec3 reflectance = is_bsdf ? vec3(1.0) : refl;
+		float alpha2 = is_bsdf ? 0.0 : a2;
```
again

devshgraphicsprogramming replied:
can support rough difftrans at no cost

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -735,179 +734,307 @@ void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
+		if (is_bsdf && is_valid)
```
this check should move up... a lot (if bsdf and not valid, you can skip all this)

devshgraphicsprogramming replied:
basically it should be
```glsl
float pdf = 0.0;
vec3 remainder = vec3(0.0);
if (cosFactor>FLT_MIN)
{
    bool is_valid = true;
    if (is_bsdf)
    {
        is_valid = //special check
    }

    if (is_valid)
    {
         //the meat of the function
    }
}
vec3 eval = remainder*pdf;
writeRegisters(pdf,eval);
```

devshgraphicsprogramming replied:
I messed up and gave bad advice

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -735,179 +734,307 @@ void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+	const vec3 refl = params_getReflectance(params);
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))//does cos>0 check even has any point here? L comes from a sample..
```
in `eval_and_pdf` cost factor check is needed to skip useless work

devshgraphicsprogramming replied:
the `cosFactor` check!

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +1055,193 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		const bool refraction = is_bsdf ? irr_glsl_partitionRandVariable(refractionProb, u.z, rcpChoiceProb) : false;
+		rcpChoiceProb = is_bsdf ? rcpChoiceProb : 1.0;
+		rem *= rcpChoiceProb;
+		rem *= refraction ? (vec3(1.0) - fr) : fr;
```
NOOOOOOOOOOOOOOOOOOOOOOOOOOOOPPPPPPPPPPPPPEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE

Conductor ONLY!

Othewise fresnel goes into the PDF!

Crisspl committed:
material compiler: Separate instruction streams for front and back (done)

Crisspl committed:
Some bug fixes

Crisspl committed:
Got rid of stuff from old twosided

Crisspl committed:
material compiler: Some optimizations

Crisspl committed:
Updated LinearAddressAllocator docs comment

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -633,14 +633,17 @@ float irr_glsl_blinn_phong_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_An
 {
-	uint op = instr_getOpcode(instr);
+	const uint op = instr_getOpcode(instr);
 
```
you forgot to clamp all the `dot` variables before using them...

Also a quick check for the **after clamp value** of `NdotL` coudl save you a lot of computation here (and data fetches)

devshgraphicsprogramming replied:
> Also a quick check for the **after clamp value** of `NdotL` coudl save you a lot of computation here (and data fetches)

Applicable to BRDF leafs only.

Only stuff that has nothing to do with the PDF computation (so all data except for roughness) and computation of Fresnel and G2.

devshgraphicsprogramming replied:
ok looks good, checked it all

devshgraphicsprogramming commented on src/irr/ext/MitsubaLoader/CMitsubaLoader.cpp:
```diff
@@ -179,42 +179,58 @@ bsdf_data_t irr_glsl_MC_fetchBSDFData(in uint ix)
+	back.offset = InstData.data[InstanceIndex].back_instr_offset;
+	back.count = InstData.data[InstanceIndex].back_rem_pdf_count;
+
+	return precomp.NdotV > 0.0 ? front : back;
```
use a proper branch... otherwise you're doing 2x the memory fetches

also its not `NdotV` that you must use (interpolated shading normals can fuck this up) .... but take in a boolean `frontFace` which in the rasterization shader would be `gl_FrontFacing`

Crisspl committed:
Some fixes regarding infinite PDF

Crisspl committed:
VT seems to work with newest changes

Crisspl committed:
A few fixes to new VT

Crisspl committed:
mitsuba loader: Changes regarding new VT

Crisspl committed:
Added callback for determining dynamically created physical storage extent per format class

Crisspl committed:
Update ex20 megatexture example

Crisspl committed:
Caching padded mip-mapped images could take too much memory

Crisspl committed:
Dont create deriv map and blendweight images more than once

Crisspl committed:
Some material compiler adjustments

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -72,10 +87,6 @@ bool instr_getWeightTexPresence(in instr_t instr)
-	return instr_get3rdParamTexPresence(instr);
-}
 
 bool instr_getTwosided(in instr_t instr)
```
whats this still doing?

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
 #ifdef GEN_CHOICE_STREAM
-void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
+// sample and microfacet are inout because of handleTwosided() only
+void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache _microfacet)
```
update comment

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -265,238 +306,125 @@ vec4 readReg4(in uint n)
+	vec3 w = params_getBlendWeight(params);
+	bxdf_eval_t bxdf1 = srcs[0].xyz;
+	bxdf_eval_t bxdf2 = srcs[1].xyz;
+	float w_pdf = (w.x + w.y + w.z) * 0.3333333;
```
wait what?
CIE XYZ both here and in the generator!

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
+		bxdf_eval_scalar_part = 0.5 * G2_over_G1 * G1_over_2NdotV * ndf_val; //cos*D*G2/(4*cos*cos)
```
why don't you just do `G2_over_G1` here and then mul the thing by `pdf` after you get it.. since the PDF is exactly the other part (`G1_over_2NdotV * ndf_val *0.5` for conductor and the other thing for dielectric refraction) ?

devshgraphicsprogramming replied:
new implementation is nice

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+			float reflectance = dot(CIE_XYZ_Luma_Y_coeffs, fr);
+			reflectance = refraction ? (1.0 - reflectance) : reflectance;
+			pdf *= reflectance;
+			bxdf_eval *= reflectance;
```
listen to me...
and only compute the remainder
https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r526500143
mul by pdf at the end!

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r526501255

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +956,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+#endif
+		pdf /= rcpChoiceProb;
+		float eta = dot(CIE_XYZ_Luma_Y_coeffs, ior[0]);
+		float rcpEta = 1.0/eta;
```
conductor need to be a blend

Crisspl replied:
???

devshgraphicsprogramming replied:
I have no idea what I meant here.

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
 	uint ix = 0u;
 	instr_t instr = irr_glsl_MC_fetchInstr(stream.offset);
 	uint op = instr_getOpcode(instr);
+	float weight = 1.0;
```
instead of weight.. use `out_pdf` right away
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r509400689

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +956,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	{} //empty braces for `else`
+
+	out_remainder = rem; // TODO not multiply rem by weight??
+	out_pdf = weight*pdf; 
```
could have been `out_pdf *= pdf`

also good to call it `localPdf` instead of `pdf`

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,203 @@
-
-#ifndef ONLY_ONE_NDF
-		} else
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
```
dont use `cos_eval_DG` .. break them down just like you did the remainder functions in the remainder stream

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-void flipCurrInteraction()
+//TODO move or rename those
+//ggx
+float irr_glsl_ggx_height_correlated_cos_eval_DG(in float NdotH, in float NdotL, in float NdotV, in float a2)
```
if you heed my advice... you'll no longer need all these
https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r526504617

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -265,238 +306,125 @@ vec4 readReg4(in uint n)
+}
+void updateCurrInteraction(in MC_precomputed_t precomp, in vec3 N)
+{
+	setCurrInteraction(precomp.NdotV > 0.0 ? N : -N, precomp.V, precomp.pos);
```
ekhm what is doing on?

Crisspl replied:
this is bug actually, precomputed normals already have correct orientation (sign)

devshgraphicsprogramming replied:
seems you dont need to `updateCurrInteraction` anymore...

devshgraphicsprogramming replied:
does function even need to exist?

devshgraphicsprogramming replied:
ah yeah bumpmaps

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -265,238 +306,125 @@ vec4 readReg4(in uint n)
+}
+void setCurrInteraction(in MC_precomputed_t precomp)
+{
+	setCurrInteraction(precomp.NdotV > 0.0 ? precomp.N : -precomp.N, precomp.V, precomp.pos);
```
and what is doing on here?

Crisspl replied:
setting initial interaction with correctly oriented normal

devshgraphicsprogramming replied:
I told you that it should depent on an input boolean (which would usually be `gl_FrontFacing` passed to this function) and not the dot product of the view vector with the normal, because the normal could have been a smooth interpolated normal!

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-		handleTwosided(ts, instr);
+		uint op = instr_getOpcode(instr);
+
+		bool skip = (i == generator_offset);
+		// skip deltas
+#ifdef OP_THINDIELECTRIC
+		skip = skip || (op == OP_THINDIELECTRIC);
+#endif
+#ifdef OP_DELTATRANS
+		skip = skip || (op == OP_DELTATRANS);
+#endif
```
also skip if roughness is really small , but I guess do that inside `instr_eval_and_pdf_execute`

Crisspl committed:
Material compiler optimizations

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+		skip = skip || (op == OP_DELTATRANS);
+#endif
+
+		if (skip)
```
pass the skip into `instr_eval_and_pdf_execute` you have divergence for doing the exact same thing (decoding registers and writing the result)

devshgraphicsprogramming replied:
and remember to init `pdf = skip ? 0.0:1.0;`

Crisspl committed:
Forgot about this

devshgraphicsprogramming commented on src/irr/ext/MitsubaLoader/CMitsubaLoader.cpp:
```diff
@@ -183,43 +179,72 @@ bsdf_data_t irr_glsl_MC_fetchBSDFData(in uint ix)
 	instr_stream_t stream;
-	stream.offset = InstData.data[InstanceIndex].bsdf_instrOffset;
-	stream.count = InstData.data[InstanceIndex].bsdf_instrCount;
+	if (precomp.NdotV > 0.0)
```
`gl_FrontFacing` not the dot

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +502,458 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-		} else
+		const float eta = colorToScalar(ior[0]);
+		const float rcp_eta = 1.0 / eta;
+		const vec3 albedo = params_getReflectance(params);
```
move albedo fetching into the `if (is_valid)`

also start off with `eta` and `rcp_eta` uninitialized.. only initialize them inside `if (is_bsdf & isTransmissionPath)¬

Crisspl replied:
it's not fetching per se, fetching is `instr_getParameters()`. Seems like parameter fetching is much less divergent than you think
Second thing: im confused, first you say to put `op_isDiffuse(op)` outside is_valid check, but you want albedo inside is_valid check

devshgraphicsprogramming replied:
> Second thing: im confused, first you say to put op_isDiffuse(op) outside is_valid check, but you want albedo inside is_valid check

My bad.. `is_valid` is only for microfacet models, diffuse only need to check that clamped cosine is `>FLT_MIN`

Then I guess albedo needs to stay outside as you've pointed out

Crisspl replied:
ok so its good

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -4,15 +4,37 @@
+bool isValidMicrofacet(in irr_glsl_IsotropicMicrofacetCache microfacet, in bool is_bsdf, in bool refraction, in float VdotL, in float eta, in float rcp_eta)
+{
+	return microfacet.NdotH >= 0.0 && !(is_bsdf && refraction && (VdotL > -min(eta, rcp_eta)));
+}
```
add this to general engine headers (call it `irr_glsl_isValidVNDFMicrofacet`)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +974,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		const float refractionProb = colorToScalar(fr);
+		float rcpChoiceProb;
+		const bool refraction = is_bsdf ? irr_glsl_partitionRandVariable(refractionProb, u.z, rcpChoiceProb) : false;
+		rcpChoiceProb = is_bsdf ? rcpChoiceProb : 1.0;
```
you should put the computation of `refractionProb`, the partitioning of the random variable and dciding refraction in the `else` clause above.

Just start with `refraction=false` and instead of screwing around with `rcpChoiceProb`, just put the subsequent `pdf /= rcpChoiceProb` in the `else` case as well

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +974,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		rcpChoiceProb = is_bsdf ? rcpChoiceProb : 1.0;
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			rem *= fr;
```
~ move this into the same `if` statement above ~

devshgraphicsprogramming replied:
actually... it ALWAYS needs to be performed (because PDF is not equal to fresnel, but monochrome of the fresnel, which are not the same thing)

so no need for the if statement

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +974,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+#endif
+		pdf /= rcpChoiceProb;
+		float eta = colorToScalar(ior[0]);
+		float rcpEta = 1.0/eta;
```
start with them undefined, only compute is you decide on doing a `refraction` in the `else` case above

devshgraphicsprogramming replied:
actually its so cheap you dont need to if-statement around it, other than putting them in the `else` (benefit from an already existing branch)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +974,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		if (op == OP_CONDUCTOR)
+			rem *= fr;
+#endif
+		pdf /= rcpChoiceProb;
```
see https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r527032102

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -265,238 +308,125 @@ vec4 readReg4(in uint n)
-#endif
+	//vec3 thickness_sigma = params_getSigmaA(params);
+	
+	vec3 ws = irr_glsl_fresnel_dielectric_frontface_only(eta, microfacet.isotropic.VdotH);
```
we expect that the COAT has already computed the fresnel for itself (dont want double application of fresnel)

so `ws` shouldnt exist and the computation at the end should be `coat+coated*wd`

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,205 @@
-		instr_execute_cos_eval_CONDUCTOR(instr, regs, bxdf_eval_scalar_part, params, bsdf_data);
-	} else
+				if (op == OP_CONDUCTOR)
+					bxdf_eval *= fr;
```
pretty sure this is wrong for a pure EVAL stream

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -16,16 +18,20 @@ void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr
 		ior2 = matrixCompMult(ior, ior);
 		params = instr_getParameters(instr, bsdf_data);
-		const float ior_scalar = dot(CIE_XYZ_Luma_Y_coeffs, ior[0]);
+		microfacet = _microfacet;
```
this should move to the second if statement... because COATING shouldn't need it (because it need not concern itself with the microfacet for specular reflection, i.e. the fresnel of the COATING part)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,205 @@
-#endif
-	}
+				const float NdotL = cosFactor;
+				const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
```
move the computation of clamped `NdotV` outside (initialize `is_valid` with `NdotV>FLT_MIN`)

the clamped `VdotH` can also be moved up and used in the valid microfacet check somewhat (but before clamp, to check that `VdotH>=0.0`)

devshgraphicsprogramming replied:
never mind the `VdotH` check, i forgot this is EVAL only

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,205 @@
+			} else
+#endif
+#if defined(OP_CONDUCTOR) || defined(OP_DIELECTRIC)
+			if (op_hasSpecular(op))
```
move this out to contain the `is_valid` check (or move the `is_valid` and `refraction` stuff into the `op_hasSpecular` if case, depending which way you look at it)... because right now you're doing a specular refraction feasibility check on the DIFFTRANS material

Crisspl committed:
Ok, it's working

Crisspl committed:
Material compiler is basically finished!

Crisspl committed:
Got rid of duplicate computation in ggx and beckmann cos_eval_and_pdf()

Crisspl committed:
Fixed some more issues

Crisspl committed:
Skip on very low alpha

Crisspl committed:
Merge branch 'mitsuba-vt2' into mitsuba

Crisspl committed:
mitsuba loader: Got rid of some not needed VT constants

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -66,19 +101,19 @@ void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr
+					bxdf_eval_scalar_part = irr_glsl_beckmann_height_correlated_cos_eval_DG_wo_clamps(NdotH2, NdotL2, NdotV2, a2);
 #else
-					bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
+					bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG_wo_clamps(NdotH2, TdotH2, BdotH2, NdotL2, TdotL2, BdotL2, NdotV2, TdotV2, BdotV2, ax, ax2, ay, ay2);
```
how do you handle `NdotV` being negative?

Crisspl replied:
abs or max obviously

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -31,28 +31,63 @@ void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr
+			const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
+#endif
+
+			const float NdotH = irr_glsl_conditionalAbsOrMax(is_bsdf, microfacet.isotropic.NdotH, 0.0);
```
no need to clamp `NdotH` by convention `H` must be in the positive hemisphere (otherwise `is_valid` will be false for bsdf and a brdf will not allow `L` and `V` outside of upper hemisphere, so half vector must also be in the hemisphere)

devshgraphicsprogramming commented on include/irr/builtin/glsl/bxdf/bsdf/specular/beckmann.glsl:
```diff
@@ -145,6 +145,34 @@ float irr_glsl_beckmann_smith_height_correlated_dielectric_cos_eval_wo_cache_val
+    );
+
+    return remainder*pdf;
+}
```
save us the torture of a big and complex wrapper around calling the `remainder_and_pdf` outselves at the target site

I'd much rather see a function in the top level header like this:
```glsl
float irr_glsl_bxdf_remainder_to_eval(in float remainder, in float pdf)
{
    return remainder*pdf;
}
vec3irr_glsl_bxdf_remainder_to_eval(in vec3 remainder, in float pdf)
{
    return remainder*pdf;
}
```

because its just 6 lines to maintain, as opposed to a separate `eval_and_pdf` boilerplate function for every BxDF in our library

Crisspl committed:
mat compiler: Resolved a few comments

Crisspl committed:
mitsuba loader/mat compiler: Seems like this commit resolves all the comments

Crisspl committed:
Go with microfacet only IR nodes

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,237 @@
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
+	const bool run = !skip;
+
+	if (run && is_bxdf_or_combiner)
```
I'd fold `positiveCosFactor` into `run` and make an `II` for `is_combiner` (so COAT and BLEND fetch the params, ior and bsdf data)

devshgraphicsprogramming replied:
just be careful because `positiveCosFactor` is now folded into `run` and unless you're able to tell if your COAT coats a DIFFTRANS or BLEND with at least one BSDF as a child, you need to fetch the data

devshgraphicsprogramming replied:
still a TODO

devshgraphicsprogramming replied:
move intitialization of `is_valid` from here
https://github.com/Crisspl/IrrlichtBAW/blob/mitsuba/include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl#L13

to here (and initialize to true) https://github.com/Crisspl/IrrlichtBAW/blob/mitsuba/include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl#L42

only use if you have `OP_DIELECTRIC` defined

(Current usage confuses me and probably the compiler as well)

devshgraphicsprogramming replied:
> I'd fold `positiveCosFactor` into `run` and make an `II` for `is_combiner` (so COAT and BLEND fetch the params, ior and bsdf data)

this is only relevant is COAT or BLEND are considered BRDFs by you (`is_bsdf` used for clamping `NdotV` and `NdotL` could be false in these cases)

that would actually prevent the loading of `bsdf_data` and `ior` under your current code when `NdotV` or `NdotL` are negative

devshgraphicsprogramming replied:
also on line https://github.com/Crisspl/IrrlichtBAW/blob/mitsuba/include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl#L62

its safe to use the `wo_clamps` variant

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,237 @@
+		{}
+	}
+
+	bxdf_eval_t result = bxdf_eval;
```
can you just do the rest on `bxdf_eval` or just call `bxdf_eval` `result` right away?

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,237 @@
-		} else
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		bool is_valid = (NdotV > FLT_MIN);
```
can't you hoist these two and fold it into the `positiveCosFactor` check?

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,237 @@
+
+#if defined(OP_DIELECTRIC) || defined(OP_CONDUCTOR)
+		is_valid = irr_glsl_isValidVNDFMicrofacet(microfacet.isotropic, is_bsdf, refraction, s.VdotL, eta, rcp_eta);
+#endif
```
a function where you don't need to compute the PDF (such as rasterization's EVAL only), does not need this check (its redundant, the BSDF transmission case does it anyway, and BRDF checks NdotV and NdotL)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,237 @@
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotH2 = microfacet.TdotH * microfacet.TdotH;
+			const float BdotH2 = microfacet.BdotH * microfacet.BdotH;
+#endif
```
why don't we precompute the squared microfacet and interaction stuff?
(obvs just for this backend, not in the general engine-wide structs)

Crisspl replied:
if we dont in general, why would we here?

devshgraphicsprogramming replied:
because here it would be useful I guess?

devshgraphicsprogramming replied:
I'm not suggesting adding fields to `irr_glsl_AnisotropicViewSurfaceInteraction ` or `irr_glsl_AnisotropicMicrofacetCache`

just creating 1 struct each, for the usage of the material compiler glsl backend which declares a the above as a member, plus the extra fields (poor man's inheritance)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl:
```diff
@@ -3,150 +3,237 @@
-	} else
-#endif
+				float VdotH = microfacet.isotropic.VdotH;
+				float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
```
`VdotH` does not need to be clamped... `VdotH<FLT_MIN` is simply an invalid microfacet normal (was checked by `is_valid` before anyway OR guaranteed by constraints on NdotV and NdotL for BRDFs)

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527890123

Crisspl committed:
Brought back mesh optimizer

Crisspl committed:
Resolved new comments

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+		if (op_isDiffuse(op))
+		{
+			eval = albedo * irr_glsl_oren_nayar_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, a2);
```
use the `wo_clamps` variant

devshgraphicsprogramming replied:
tied up with this https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r528997756

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+			const float TdotH2 = microfacet.TdotH * microfacet.TdotH;
+			const float BdotH2 = microfacet.BdotH * microfacet.BdotH;
+			const float NdotH2 = microfacet.isotropic.NdotH2;
+			const float NdotL = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
```
you could hoist this above (the computation only) and put an extra `if (NdotL>FLT_MIN)` inside the diffuse case (generator for a BRDF cannot actually generate a light direction below hemisphere, so all good) this will also allow you to use the `wo_clamps` variant

devshgraphicsprogramming replied:
put the extra `NdotL>FLT_MIN` check inside

Crisspl replied:
i cant, keep getting nans then

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -265,238 +324,125 @@ vec4 readReg4(in uint n)
-#endif
+	// TODO include thickness_sigma in diffuse weight computation: exp(sigma_thickness * freePath)
+	// freePath = ( sqrt(refract_compute_NdotT2(NdotL2, rcpOrientedEta2)) + sqrt(refract_compute_NdotT2(NdotV2, rcpOrientedEta2)) )
+	vec3 fresnelNdotV = irr_glsl_fresnel_dielectric_frontface_only(eta, currInteraction.isotropic.NdotV);
```
careful about `NdotV` being negative (interpolated normals)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+				reflectance = refraction ? (1.0 - reflectance) : reflectance;
+				pdf *= reflectance;
+				eval *= reflectance;
+			}
```
a suggestion, to save on 3x RGA mult and branches (Also your current code seems to be wrong)
```glsl
#ifndef NO_BSDF
			if (is_bsdf)
			{
				float LdotH = microfacet.isotropic.LdotH;
				float VdotHLdotH = microfacet.isotropic.VdotH * LdotH;
				LdotH = abs(LdotH);
#ifdef NDF_GGX
				if (ndf == NDF_GGX)
					eval_scalar_part = irr_glsl_ggx_microfacet_to_light_measure_transform(eval_scalar_part, NdotL, refraction, VdotH, LdotH, VdotHLdotH, eta);
				else
#endif
					eval_scalar_part = irr_glsl_microfacet_to_light_measure_transform(eval_scalar_part, NdotV, refraction, VdotH, LdotH, VdotHLdotH, eta);

				float reflectance = colorToScalar(fr);
				reflectance = refraction ? (1.0 - reflectance) : reflectance;
				pdf *= reflectance;
			}
#endif 

			eval = fr*eval_scalar_part;
```

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-	if (!ts_flag && instr_getTwosided(instr))
+
+	vec3 eval = vec3(0.0);
+	float pdf = 0.0;
```
init with `skip ? 0.0:1.0` otherwise pdf always 0

https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r526974227

devshgraphicsprogramming replied:
ok missed the assignment

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+	// (is_bxdf && is_bsdf) -> BSDF
+	// (is_bxdf && !is_bsdf) -> BRDF
+	const bool is_bxdf_or_combiner = op_isBXDForCoatOrBlend(op);
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
```
unused variable, delete

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+		const float rcp_eta = 1.0 / eta;
+		const vec3 albedo = params_getReflectance(params);
+
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
```
this one you can hoist and add the condition `>FLT_MIN` to `is_bxdf && run` because VNDF sampling (VNDF sampling can produce L samples that are shadowed, but cannot produce V samples that are masked)

devshgraphicsprogramming replied:
and obvs initialize `is_valid=true` after the hoist

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-	if (op==OP_PLASTIC) {
-		pdf = instr_execute_cos_eval_pdf_PLASTIC(instr, regs, bxdf_eval_scalar_part, params, bsdf_data, s, pdf);
-	} else
+				fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH);
```
you probably want to convert `ior` to a scalar value (use `eta*eta` I guess?)

Crisspl replied:
no, why?

devshgraphicsprogramming replied:
this is why https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r530488943

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-		instr_execute_cos_eval_DIELECTRIC(instr, regs, params, bsdf_data);
-	} else
+			{
+				float reflectance = colorToScalar(fr);
```
if you do `https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r529025362` then you can use any of the `reflectance` channels

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r530488943

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-	uint op = instr_getOpcode(instr);
+#if defined(OP_DIELECTRIC) || defined(OP_CONDUCTOR)
+		is_valid = irr_glsl_isValidVNDFMicrofacet(microfacet.isotropic, is_bsdf, refraction, s.VdotL, eta, rcp_eta);
+#endif
```
I wonder if you could push all this (and the rcp eta and eta computation) INTO the `else` case of `op_isDiffuse`

obvs after `NdotV>FLT_MIN` hoist ( https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r529019954 )

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+			const float TdotH2 = microfacet.TdotH * microfacet.TdotH;
+			const float BdotH2 = microfacet.BdotH * microfacet.BdotH;
+			const float NdotH2 = microfacet.isotropic.NdotH2;
```
same comment as https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r528886851

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
-	irr_glsl_BSDFSample s = irr_bsdf_cos_generate(gcs, rand);
-	irr_glsl_updateBxDFParams(currBSDFParams, s, currInteraction);
-	vec3 rem = irr_bsdf_cos_remainder_and_pdf(rnps, s, pdf);
+	instr_t generator;
```
unused variable

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	uint generator_rnpOffset;
+	bool invalid_microfacet;
+	irr_glsl_LightSample s = irr_bsdf_cos_generate(precomp, gcs, rand, generator_rem, generator_pdf, microfacet, generator_rnpOffset, invalid_microfacet);
```
get rid of `invalid_microfacet`, a generator cannot generate such

devshgraphicsprogramming replied:
its unused var anyway

Crisspl committed:
Matt's snippet

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
 #endif //OP_BLEND
+#ifdef OP_COATING
+		if (op==OP_COATING) {
+			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
```
hoist this fetch, both BLEND and COATING branch do it

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+
+			uint coated_ix = instr_getRightJump(instr);
+			ix = choseCoated ? coated_ix:(ix+1u);
+			out_pdf /= rcpChoiceProb;
```
The COAT and BLEND only diverge in how they compute the RGB weight

after you get the weight, the rest of the operations are the same (RGB to monochrome, partitioning, pdf calc, jump calc)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	if (op == OP_DELTATRANS)
+	{
+		s = irr_glsl_createLightSample(-precomp.V, -1.0, currInteraction.T, currInteraction.B, currInteraction.isotropic.N);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
```
there's no point to computing this, since it will be a tranmission interaction, ergo the `out_microfacet` which is only for reflection wont be used at all

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -688,18 +688,25 @@ void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, i
+					eval_scalar_part = irr_glsl_ggx_microfacet_to_light_measure_transform(eval_scalar_part, NdotL, refraction, VdotH, LdotH, VdotHLdotH, eta);
+				else
+#endif
+					eval_scalar_part = irr_glsl_microfacet_to_light_measure_transform(eval_scalar_part, NdotV, refraction, VdotH, LdotH, VdotHLdotH, eta);
```
the top bit of code repeats this

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	if (op == OP_THINDIELECTRIC)
+	{
+		const vec3 luminosityContributionHint = CIE_XYZ_Luma_Y_coeffs;
+		s = irr_glsl_thin_smooth_dielectric_cos_generate(currInteraction, u, ior2[0], luminosityContributionHint);
```
the most complicated form of the function actually cuts down on computation
https://github.com/Devsh-Graphics-Programming/Nabla/blob/881deedeac39b6141d061756fa9b00951fb57f28/include/irr/builtin/glsl/bxdf/bsdf/specular/dielectric.glsl#L17

returns remainder metadata, which you can put into this
https://github.com/Devsh-Graphics-Programming/Nabla/blob/881deedeac39b6141d061756fa9b00951fb57f28/include/irr/builtin/glsl/bxdf/bsdf/specular/dielectric.glsl#L48

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+		if (op==OP_COATING) {
+			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+			vec3 eta = bsdf_data_decodeIoR(bsdf_data, OP_COATING)[0];
+			vec3 fresnel = irr_glsl_fresnel_dielectric_frontface_only(eta, currInteraction.isotropic.NdotV);
```
careful of negative `currInteraction.isotropic.NdotV` 

clamp

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		s = irr_glsl_thin_smooth_dielectric_cos_generate(currInteraction, u, ior2[0], luminosityContributionHint);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		rem = irr_glsl_thin_smooth_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ior2[0], luminosityContributionHint);
+	} else
```
every case after this (diffuse and microfacet) needs to check for `clampedOrAbsNdotV>FLT_MIN` and set remainder+pdf to 0.0 if it doesn't pass

devshgraphicsprogramming replied:
its not only an optimization, its necessary

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		const float BdotH2 = out_microfacet.BdotH * out_microfacet.BdotH;
+		const float NdotH2 = out_microfacet.isotropic.NdotH2;
+		const float NdotL = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
```
Hoist...

https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r529049195

devshgraphicsprogramming replied:
obviously no point hoisting `NdotL`

only hoist `NdotV`

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	{
+		const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+		const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
+		const float NdotV2 = currInteraction.isotropic.NdotV_squared;
```
similar comment to https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r528886851

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		else
+#endif
+		{
+			fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH);
```
beware of negative `VdotH`

`NdotH` can't be negative no matter BRDF or BSDF
but `VdotH` can be negative for a BSDF (even from generator)

devshgraphicsprogramming replied:
also in order to benefit from skipping a mul by RGB fresnel followed by divide by monochrome fresnel of the remainder, the fresnel function used needs to be monochrome (not RGB)

devshgraphicsprogramming replied:
basically use the `eta` and `rcpEta` you're computing just outside and after this branch ;)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+		{
+			fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH);
```
prevent negative `VdotH` for BRDFs https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r529051069

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+
+		const float TdotH2 = out_microfacet.TdotH * out_microfacet.TdotH;
+		const float BdotH2 = out_microfacet.BdotH * out_microfacet.BdotH;
+		const float NdotH2 = out_microfacet.isotropic.NdotH2;
```
same comment as https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r528886851

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		END_CASES
+
+		rem *= G2_over_G1;
+		pdf *= irr_glsl_smith_VNDF_pdf_wo_clamps(ndf_val, G1_over_2NdotV);
```
the PDF looks suspicious, you're using the VNDF to reflective differential factor transform here....

you should be using the one that can account for refraction
https://github.com/Devsh-Graphics-Programming/Nabla/blob/shader_pipeline_irrlichtbaw/include/irr/builtin/glsl/bxdf/geom/smith/common.glsl#L33

Crisspl committed:
mat compiler: Fix pdf to account for transmission

Crisspl committed:
ggx and beckmann cos_remainder_and_pdf() changes

Crisspl committed:
Comments resolutions

Crisspl committed:
ex22 rtao little update

not much but i inserted some useful comments

Crisspl committed:
material compiler: Precomputed V and H squared dots with tangents

Known bug: some of the emitters in staircase scene are black because they have normals wrong way around or vertex order inverted (so they are recognized as backfacing) and so NdotV is negative in irr_bsdf_cos_generate()

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +853,207 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+			else
+#endif
+			{
+				fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
```
use monochrome dielectric fresnel function, if you want to assume that fresnel in the remainder cancels out with the reflection/refraction probability in the PDF

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r529052459

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +853,207 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	const bool positiveNdotV = (NdotV > FLT_MIN);
 
+	float localPdf = positiveNdotV ? 1.0 : 0.0;
+	vec3 rem = positiveNdotV ? vec3(1.0) : vec3(0.0);
```
just initialize to 0.0, overwrite in the `if (positiveNdotV)` case/opcodes

devshgraphicsprogramming replied:
swap out some of the `*=` for `=`

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +853,207 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+			const float LdotH = out_microfacet.inner.isotropic.LdotH;
+			const float VdotHLdotH = VdotH * LdotH;
+			// a trick, pdf was already multiplied by transmission/reflection choice probability above
+			const float reflectance = refraction ? 0.0 : 1.0;
```
then make a specialized `irr_glsl_smith_VNDF_pdf_wo_clamps` that transforms a `FVNDF`  and does not take a `refraction` parameter ;)

devshgraphicsprogramming replied:
```glsl
float irr_glsl_smith_FVNDF_pdf_wo_clamps(in float fresnel_ndf, in float G1_over_2NdotV, in float absNdotV, in bool transmitted, in float VdotH, in float LdotH, in float VdotHLdotH, in float orientedEta)
{
    float FNG = fresnel_ndf*G1_over_2NdotV;
    float factor = 0.5;
    if (transmitted)
    {
        const float VdotH_etaLdotH = (VdotH+orientedEta*LdotH);
        // VdotHLdotH is negative under transmission, so this factor is negative
        factor *= -2.0*VdotHLdotH/(VdotH_etaLdotH*VdotH_etaLdotH);
    }
    return FNG*factor;
}
```

and then rewrite https://github.com/Devsh-Graphics-Programming/Nabla/blob/shader_pipeline_irrlichtbaw/include/irr/builtin/glsl/bxdf/geom/smith/common.glsl#L33

in terms of the new thing

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -529,385 +477,362 @@ vec3 fetchTex(in uvec3 texid, in vec2 uv, in mat2 dUV)
-			currBSDFParams.BdotH = -currBSDFParams.BdotH;
+			eval = albedo * irr_glsl_oren_nayar_cos_remainder_and_pdf_wo_clamps(pdf, a2, s.VdotL, NdotL, NdotV);
+			pdf *= is_bsdf ? 0.5 : 1.0;
+			eval *= pdf;
```
put another if-statement inside to check for `NdotL >FLT_MIN`

https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r528997756

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -529,385 +477,362 @@ vec3 fetchTex(in uvec3 texid, in vec2 uv, in mat2 dUV)
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
+				const float VdotH = is_bsdf ? microfacet.inner.isotropic.VdotH : max(microfacet.inner.isotropic.VdotH, 0.0);
```
just use `abs(microfacet.inner.isotropic.VdotH)`, if you have a BRDF the smith masking and shadowing function will kill the wrongly computed fresnel (which would compute to 100% anyway with VdotH clamped to 0)

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -529,385 +477,362 @@ vec3 fetchTex(in uvec3 texid, in vec2 uv, in mat2 dUV)
-	if (op==OP_DIFFTRANS) {
-		instr_execute_cos_eval_DIFFTRANS(instr, regs, params, bsdf_data);
-	} else
+					fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH);
```
monochrome fresnel, use `eta` and `rcpEta`

Crisspl committed:
Added material compiler GLSL to mitsuba loader global metadata

Crisspl committed:
rtao example to new api modifications

Crisspl committed:
yet another comments resolved

devshgraphicsprogramming commented on examples_tests/22.RaytracedAO/mesh.frag:
```diff
@@ -4,12 +4,13 @@ layout(location = 3) uniform vec3 color;
 layout(location = 4) uniform uint nasty;
 layout(location = 5) uniform uint lightID;
 
+//TODO not needed any more, right?
```
yes its not needed

devshgraphicsprogramming commented on examples_tests/22.RaytracedAO/mesh.frag:
```diff
@@ -4,12 +4,13 @@ layout(location = 3) uniform vec3 color;
 in vec2 uv;
 in vec3 Normal;
 
-layout(location = 0) out vec3 pixelColor;
+layout(location = 0) out vec3 pixelColor; //TODO i think this one we will skip in gbuffer fill pass?
 layout(location = 1) out vec2 encodedNormal;
 layout(location = 2) out uint lightIndex;
```
ouput:
- meshbuffer ID + 1bit for frontfacing vs. back facing
- triangle ID (dont use yet obvs)
- encoded normal
- uv coord (in full range of a sfloat)


From meshbuffer/object ID you shall be able to get access to all data about shading model (instruction streams) and light emittance

devshgraphicsprogramming commented on include/irr/builtin/material_compiler/glsl/common.glsl:
```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+	if (op_hasSpecular(op))
+	{
+		const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+		const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
```
why are you not using the procmputed squared values  from your interaction struct?

Crisspl replied:
you're looking at old code

devshgraphicsprogramming replied:
darn github

Crisspl committed:
Resolved last comments in mat compiler
